### ISSUE TITLE
Neo4j module support

### ISSUE THREAD

**Issue by rickardoberg** [neo4j/neo4j#12952]
Neo4j 4.x, embedded

It is currently impossible to use Neo4j 4.x in an embedded application, if you want to use the Java module system. The reason is that there are too many packages that are split between jar files, such as:
org.neo4j.dbms.api,org.neo4j.dbms.database,org.neo4j.dbms.identity,org.neo4j.dbms.systemgraph:
neo4j-4.x.x.jar and neo4j-graphdb-api-4.x.x.jar
org.neo4j.kernel.api,org.neo4j.kernel.impl.index.schema:
neo4j-kernel-4.x.x.jar and neo4j-kernel-api-4.x.x.jar
and so on

I'm sure you're aware of this, so the questions are:
1) Any known workarounds for embedded apps until this is fixed?
2) Any ideas on when this will be fixed? (release nr or date)

Thanks!

**Comment by rickardoberg** [comment]
In the meantime, we have had to create our own shaded jar with Neo4j+Lucene+Shiro+Scala libraries (all of which were problematic from module point of view). Here's the POM to do it, for anyone else having these issues:
https://github.com/Cantara/neo4j-apoc-jpms-shaded

**Comment by rickardoberg** [comment]
I tried with Neo4j 5.1.0, and same problem. The jars now use Automatic-Module-Name, which is great, but there are split packages everywhere, so still unusable, forcing us to create a shaded jar with all the classes combined. Disappointed that this could not be fixed for 5.x, but hopefully by 6.x this might get fixed?

**Comment by klaren** [comment]
Hello @rickardoberg, 

This is known and has not yet reached the top of the priority list. The only workaround I can think of is to shade everything into an uber-jar to eliminate the duplicate exports. Unfortunately, I cannot give you any estimate on when this will be prioritized, there are way too many variables involved.

**Comment by rickardoberg** [comment]
It is now June of 2025, and no effort has been made (apparently) to fix this. The current release (2025.05.0) still has the same split package issues. There seems to be no interest in having artifacts with clean separation between them, or any notion of internal/API/SPI. 

So, given this sad state of affairs I'm going to suggest the same thing I said to APOC: just give up and release an uberjar with everything yourself. That way we-the-users don't have to do it at least.

**Comment by rickardoberg** [comment]
Ping @klaren

**Comment by klaren** [comment]
@rickardoberg 
You are correct that there is still no module support. It hasn't been prioritized because less than 0.5% of users embed Neo4j, with the vast majority using the server installation. Additionally, there hasn't been a clear benefit internally for us to introduce modules, making it difficult to justify. I agree it's unfortunate, but that's the current reality.

I will raise the request to add an all-in-one jar file to Maven Central to make it easier for you.

**Comment by rickardoberg** [comment]
@klaren great! thank you so much!
